server.go

```diff
@@ -379,18 +431,54 @@ func Serve(opts *ServeConfig) {
 		}
 	}()
 
-	// Set our new out, err
+	// Set our stdout, stderr to the stdio stream that clients can retrieve
+	// using ClientConfig.SyncStdout/err.
+	//
+	// In test mode, we use a multiwriter so that the data continues going
+	// to the normal stdout/stderr so output can show up in test logs. We
+	// also send to the stdio stream so that clients can continue working
+	// if they depend on that.
+	if opts.Test != nil {
+		// In test mode we need to maintain the original values so we can
+		// reset it.
+		defer func(out, err *os.File) {
```

Comment:
This is racy. Because os.Stdout is essentially global mutable state, if multiple servers are run in the same process in parallel (which happens in tests with some frequency), there's a good chance the following will happen:

* Server 1 sets os.Stdout to stdout_w, stores /dev/stdout in the defer
* Server 2 sets os.Stdout to stdout_w, stores stdout_w from Server 1 in the defer
* Server 1 finishes and the defer triggers, os.Stdout is now /dev/stdout
* Server 2 finishes and the defer triggers, os.Stdout is now stdout_w from Server 1

What's worse, once it happens, it will never be corrected; each new server will assume that os.Stdout is set to /dev/stdout and will reset it faithfully, so once a wrong value is persisted, it's persisted indefinitely. Even with a moderate number of tests, this race condition has a good chance of triggering. For large providers with lots of tests, it's almost certain to trigger.

I've pushed the branch `paddy-test-mode-race` with an update the test mode test to demonstrate this race condition.

Reply:
Ah yes, this was designed on purpose assuming only one test was started at a time. Let's chat about your use case here and how multiple start at once.

Reply:
I think I have a solution to this, you now opt-in to stdio syncing with `SyncStdio` (in test config). This defaults to false since its messy/complex, which is the behavior you probably want.

Reply:
(I know you fixed this, but to document it: the use case where multiple start at a time is when you use a server in multiple tests, and Go runs tests in parallel, there's no guarantee those tests won't run at the same time.)